pkg/controllers/securitypolicy_associated_controller.go

```diff
@@ -0,0 +1,241 @@
+// When a pod's label is changed.
+// When a namespace's label is changed.
+// In summary, we could roughly think if the port name of security policy exists in the
+// new pod or old pod, we should reconcile the security policy.
+
+type EnqueueRequestForNamespace struct {
+	Client client.Client
+}
+
+func (e *EnqueueRequestForNamespace) Create(_ event.CreateEvent, _ workqueue.RateLimitingInterface) {
+	log.V(1).Info("namespace create event, do nothing")
+}
+
+func (e *EnqueueRequestForNamespace) Delete(_ event.DeleteEvent, _ workqueue.RateLimitingInterface) {
+	log.V(1).Info("namespace delete event, do nothing")
+}
+
+func (e *EnqueueRequestForNamespace) Generic(_ event.GenericEvent, _ workqueue.RateLimitingInterface) {
+	log.V(1).Info("namespace generic event, do nothing")
+}
+
+func (e *EnqueueRequestForNamespace) Update(updateEvent event.UpdateEvent, l workqueue.RateLimitingInterface) {
+	obj := updateEvent.ObjectNew.(*v1.Namespace)
+
+	podList := &v1.PodList{}
+	err := e.Client.List(context.Background(), podList, client.InNamespace(obj.Name))
+	if err != nil {
+		log.Error(err, "Failed to list pod list CR")
+		return
+	}
+
+	shouldReconcile := false
+	for _, pod := range podList.Items {
+		if checkPod(&pod, "update") {
+			shouldReconcile = true
+			break
+		}
+	}
+	if !shouldReconcile {
+		log.Info("no pod in namespace is relevant", "namespace", obj.Namespace, "name", obj.Name)
+		return
+	}
+
+	err = reconcileSecurityPolicy(e.Client, podList.Items, l)
+	if err != nil {
+		log.Error(err, "Failed to reconcile security policy")
+	}
+}
+
+type EnqueueRequestForPod struct {
+	Client client.Client
+}
+
+func (e *EnqueueRequestForPod) Create(createEvent event.CreateEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(createEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Update(updateEvent event.UpdateEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(updateEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Delete(deleteEvent event.DeleteEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(deleteEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Generic(genericEvent event.GenericEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(genericEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Raw(evt interface{}, q workqueue.RateLimitingInterface) {
+	var pods []v1.Pod
+	var obj client.Object
+	var oldObj client.Object
+
+	switch et := evt.(type) {
+	case event.CreateEvent:
+		obj = et.Object.(*v1.Pod)
+	case event.UpdateEvent:
+		obj = et.ObjectNew.(*v1.Pod)
+		oldObj = et.ObjectOld.(*v1.Pod)
+		pods = append(pods, *oldObj.(*v1.Pod))
+	case event.DeleteEvent:
+		obj = et.Object.(*v1.Pod)
+	case event.GenericEvent:
+		obj = et.Object.(*v1.Pod)
+	default:
+		log.Error(nil, "Unknown event type", "event", evt)
+	}
+	pod := obj.(*v1.Pod)
+	pods = append(pods, *pod)
+	err := reconcileSecurityPolicy(e.Client, pods, q)
+	if err != nil {
+		log.Error(err, "Failed to reconcile security policy")
+	}
+}
+
+func getAllPodPortNames(pods []v1.Pod) sets.String {
```

Comment:
This function could expose potential performance issue given in large scale env. Since worse time complexity is O(N^3), it will cost much time to finish if there are multiple new pods with named ports to created.  In NCP, we can use fair iterator to avoid long time execution of one function in a thread. Here,  the situation could be different. We can think in the future there is some way to optimize pods list to boost performance. For example, using go routine to do it in parallel.

Reply:
ok